components/modal/frame.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, compose, createRef } from '@wordpress/element';
+import { ESCAPE } from '@wordpress/keycodes';
+import { focus } from '@wordpress/dom';
+
+/**
+ * External dependencies
+ */
+import clickOutside from 'react-click-outside';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import withFocusReturn from '../higher-order/with-focus-return';
+import withConstrainedTabbing from '../higher-order/with-constrained-tabbing';
+import withGlobalEvents from '../higher-order/with-global-events';
+
+class ModalFrame extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.containerRef = createRef();
+		this.handleKeyDown = this.handleKeyDown.bind( this );
+		this.handleClickOutside = this.handleClickOutside.bind( this );
+		this.focusFirstTabbable = this.focusFirstTabbable.bind( this );
+	}
+
+	/**
+	 * Focuses the first tabbable element when props.focusOnMount is true.
+	 */
+	componentDidMount() {
+		// Focus on mount
+		if ( this.props.focusOnMount ) {
+			this.focusFirstTabbable();
+		}
+	}
+
+	/**
+	 * Focuses the first tabbable element.
+	 */
+	focusFirstTabbable() {
+		const tabbables = focus.tabbable.find( this.containerRef.current );
+		if ( tabbables.length ) {
+			tabbables[ 0 ].focus();
+		}
+	}
+
+	/**
+	 * Callback function called when clicked outside the modal.
+	 *
+	 * @param {Object} event Mouse click event.
+	 */
+	handleClickOutside( event ) {
+		if ( this.props.shouldCloseOnClickOutside ) {
+			this.onRequestClose( event );
+		}
+	}
+
+	/**
+	 * Callback function called when a key is pressed.
+	 *
+	 * @param {Object} event Key down event.
```

Comment:
Super nitpick: We can be more specific about the interface of the object we're receiving, since we're using an assumed `keyCode`, it could be better to assign this type as [`KeyboardEvent`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent) which is responsible for defining the `keyCode` property.